AsyncDisplayKit/Details/ASDataController.mm

```diff
@@ -218,309 +232,137 @@ - (void)_batchLayoutAndInsertNodesFromContexts:(NSArray<ASIndexedNodeContext *>
-          [_delegate dataControllerWillDeleteAllData:self];
-        }];
-        
-        // deleteSections:
-        // Remove everything that existed before the reload, now that we're ready to insert replacements
-        NSUInteger oldSectionCount = [_editingNodes[ASDataControllerRowNodeKind] count];
-        if (oldSectionCount) {
-          NSIndexSet *indexSet = [[NSIndexSet alloc] initWithIndexesInRange:NSMakeRange(0, oldSectionCount)];
-          [self _deleteSectionsAtIndexSet:indexSet withAnimationOptions:animationOptions];
-        }
-        
-        [self willReloadDataWithSectionCount:sectionCount];
-        
-        // insertSections:
-        NSMutableArray *sections = [NSMutableArray arrayWithCapacity:sectionCount];
-        for (int i = 0; i < sectionCount; i++) {
-          [sections addObject:[[NSMutableArray alloc] init]];
-        }
-        [self _insertSections:sections atIndexSet:sectionIndexes withAnimationOptions:animationOptions];
-      }
-      
-      // insertItemsAtIndexPaths:
-      [self _insertNodes:nodes atIndexPaths:indexPaths withAnimationOptions:animationOptions];
-      
-      if (isFirstBatch) {
-        // -endUpdates
-        [_mainSerialQueue performBlockOnMainThread:^{
-          [_delegate dataController:self endUpdatesAnimated:NO completion:nil];
-        }];
-        isFirstBatch = NO;
-      }
-    }];
-    
-    if (completion) {
-      [_mainSerialQueue performBlockOnMainThread:completion];
-    }
-  });
-  if (synchronously) {
-    [self waitUntilAllUpdatesAreCommitted];
+  if (sections.count == 0) {
+    return;
   }
-}
-
-- (void)waitUntilAllUpdatesAreCommitted
-{
-  ASDisplayNodeAssertMainThread();
   
-  dispatch_group_wait(_editingTransactionGroup, DISPATCH_TIME_FOREVER);
-  
-  // Schedule block in main serial queue to wait until all operations are finished that are
-  // where scheduled while waiting for the _editingTransactionQueue to finish
-  [_mainSerialQueue performBlockOnMainThread:^{ }];
+  NSArray<NSIndexPath *> *indexPaths = [self _allIndexPathsForNodeContextsOfKind:kind inSections:sections];
+  [self _insertNodeContextsOfKind:kind atIndexPaths:indexPaths environment:environment];
 }
 
-#pragma mark - Data Source Access (Calling _dataSource)
-
 /**
- * Fetches row contexts for the provided sections from the data source.
+ * Inserts new node contexts of a certain kind at some index paths
+ *
+ * @param kind The kind of the node contexts, e.g ASDataControllerRowNodeKind
+ * @param indexPaths The index paths at which new contexts should be populated
+ * @param environment The trait environment needed to initialize contexts
  */
-- (NSArray<ASIndexedNodeContext *> *)_populateNodeContextsFromDataSourceForSections:(NSIndexSet *)sections
+- (void)_insertNodeContextsOfKind:(NSString *)kind
+                     atIndexPaths:(NSArray<NSIndexPath *> *)indexPaths
+                      environment:(id<ASTraitEnvironment>)environment
```

Comment:
Nice! Love having the environment as a parameter here so we can keep track of it. In the future we should be sure that when the environment changes, we go through all affected contexts and update them. I know @rcancro was dealing with environment changes mid-update and was super frustrated.